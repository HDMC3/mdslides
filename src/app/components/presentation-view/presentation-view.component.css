.view-container {
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden;
}

.slides-track {
    display: flex;
    width: 100%;
    height: 100%;
    transition: none;
}

.slide {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    flex-shrink: 0;
}

.next, .previous {
    position: absolute;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 75px;
    height: 100%;
    opacity: 1;
    border: none;
    background-color: var(--color-info-transparent-200);
    transition: opacity 0.3s;
}

.next {
    right: 0;
    border-left: 2px solid var(--color-primary-400);
}

.previous {
    left: 0;
    border-right: 2px solid var(--color-primary-400);
}

.next:hover, .previous:hover {
    opacity: 1;
}

.next:active, .previous:active {
    opacity: 0.2;
}

.hidde-action-buttons {
    opacity: 0;
}