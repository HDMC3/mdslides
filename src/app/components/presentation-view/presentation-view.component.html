<div class="view-container" >
    <div #slidesTrack class="slides-track">
        <div *ngFor="let slide of presentation?.slides" class="slide"></div>
    </div>
    <button [ngClass]="{'hidde-action-buttons': hiddeActionButtons}" class="next" (click)="next()">
        <nb-icon style="color: var(--color-primary-400); width: 50px; height: 50px;" icon="arrow-forward"></nb-icon>
    </button>
    <button [ngClass]="{'hidde-action-buttons': hiddeActionButtons}" class="previous" (click)="previous()">
        <nb-icon style="color: var(--color-primary-400); width: 50px; height: 50px;" icon="arrow-back"></nb-icon>
    </button>
</div>