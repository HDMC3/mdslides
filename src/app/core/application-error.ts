export class ApplicationError extends Error {
    constructor(message: string, public title?: string) {
        super(message);
    }
}
