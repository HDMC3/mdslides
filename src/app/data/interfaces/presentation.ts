import { Slide } from './slide';

export interface Presentation {
    id: string;
    title: string;
    description: string;
    slides: Slide[];
    creation_date: Date;
    modification_date: Date;
}
