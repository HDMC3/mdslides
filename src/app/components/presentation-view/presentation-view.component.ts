import { Component, ElementRef, OnInit, AfterViewInit, ViewChild, HostListener } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { MarpitService } from 'src/app/core/services/marpit.service';
import { PresentationService } from 'src/app/core/services/presentation.service';
import { Presentation } from 'src/app/data/interfaces/presentation';
import { Slide } from 'src/app/data/interfaces/slide';

@Component({
    selector: 'app-presentation-view',
    templateUrl: './presentation-view.component.html',
    styleUrls: ['./presentation-view.component.css']
})
export class PresentationViewComponent implements OnInit, AfterViewInit {

    @ViewChild('slidesTrack') slidesTrack?: ElementRef<HTMLDivElement>;
    isActive: boolean;
    presentation?: Presentation;
    offsetSlidesTrack: number;
    currentSlide?: Slide;
    hiddeActionButtons: boolean;

    constructor(
        private elementRef: ElementRef<HTMLElement>,
        private router: Router,
        private activatedRoute: ActivatedRoute,
        private presentationService: PresentationService,
        private marpitService: MarpitService
    ) {
        this.isActive = false;
        this.offsetSlidesTrack = 0;
        this.hiddeActionButtons = false;
    }

    ngOnInit(): void {

        this.elementRef.nativeElement.addEventListener('fullscreenchange', (e: any) => {
            if (this.isActive) {
                this.activatedRoute.parent?.params.subscribe(params => {
                    this.router.navigate(['/']);
                });
            }

            if (!this.isActive) this.isActive = true;
        });

        this.presentationService.presentation$.subscribe(presentation => {
            this.presentation = presentation;
        });

        this.presentationService.currentSlide$.subscribe(slide => {
            this.currentSlide = slide;
        });
    }

    ngAfterViewInit(): void {
        this.init();
        setTimeout(() => {
            this.hiddeActionButtons = true;
        }, 1000);
    }

    @HostListener('window:deviceorientation')
    onDeviceOrientation() {
        this.renderSlides();
    }

    async init() {
        try {
            await this.elementRef.nativeElement.requestFullscreen();
        } catch (error) {
            this.router.navigate(['/']);
        }
        this.renderSlides();
    }

    next = () => {
        if (!this.presentation) throw new Error();
        const totalOffset = -this.presentation.slides.length * window.innerWidth;

        if (this.offsetSlidesTrack <= totalOffset) {
            if (document.fullscreenElement) document.exitFullscreen();
            return;
        }

        if (!this.slidesTrack) throw new Error();
        this.offsetSlidesTrack -= window.innerWidth;
        this.slidesTrack.nativeElement.style.transform = `translateX(${this.offsetSlidesTrack}px)`;
    };

    previous = () => {
        if (this.offsetSlidesTrack >= 0) return;

        if (!this.slidesTrack) throw new Error();
        this.offsetSlidesTrack += window.innerWidth;
        this.slidesTrack.nativeElement.style.transform = `translateX(${this.offsetSlidesTrack}px)`;
    };

    @HostListener('document:keyup', ['$event'])
    onComponentKeyup(e: KeyboardEvent) {
        if (e.code === 'ArrowLeft') {
            this.previous();
        }
        if (e.code === 'ArrowRight') {
            this.next();
        }
    }

    renderSlides() {
        if (!this.slidesTrack || !this.presentation) throw new Error();

        const slidesElements = this.slidesTrack.nativeElement.querySelectorAll<HTMLElement>('.slide');
        let currentSlideIndex = 0;
        for (let i = 0; i < this.presentation.slides.length; i++) {
            const slide = this.presentation.slides[i];
            const { html, css } = this.marpitService.render(slide.code);
            const slideElement = slidesElements[i];
            slideElement.style.setProperty('--scale', '1');
            const shadowRoot = slideElement.shadowRoot ?? slideElement.attachShadow({ mode: 'open' });
            shadowRoot.innerHTML = `<style>${css}\n.marpit section { transform: scale(var(--scale));}</style>\n${html}`;
            this.slidesTrack.nativeElement.appendChild(slideElement);
            this.scaleSlide(slideElement);
            if (slide === this.currentSlide) {
                currentSlideIndex = i;
            }
        }

        this.offsetSlidesTrack = -currentSlideIndex * window.innerWidth;
        this.slidesTrack.nativeElement.style.transform = `translateX(${this.offsetSlidesTrack}px)`;

        setTimeout(() => {
            this.slidesTrack?.nativeElement.style.setProperty('transition', 'transform 0.2s');
        }, 0);
    }

    scaleSlide(slideElement: HTMLElement) {
        const marpitSlide = slideElement.shadowRoot?.querySelector('section');
        if (!marpitSlide) return;

        const marpitSlideWidth = marpitSlide.getBoundingClientRect().width;
        const marpitSlideHeight = marpitSlide.getBoundingClientRect().height;

        const scale = Math.min(
            window.innerWidth / marpitSlideWidth,
            window.innerHeight / marpitSlideHeight
        );

        marpitSlide.style.setProperty('--scale', `${scale}`);
    }
}
