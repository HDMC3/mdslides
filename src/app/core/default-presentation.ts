import { v4 } from 'uuid';
import { Presentation } from '../data/interfaces/presentation';

export const getDefaultPresentation = (id: string): Presentation => {
    return {
        id,
        title: 'Titulo de presentacion',
        creation_date: new Date(),
        modification_date: new Date(),
        description: '',
        slides: [
            {
                id: v4(),
                name: 'Presentacion inicial',
                code: [
                    '---',
                    'style: |',
                    'section {',
                    '   display: flex;',
                    '   flex-direction: column;',
                    '   align-items: center;',
                    '   justify-content: center;',
                    '}',
                    '---',
                    '',
                    '# Inicio de la presentacion',
                    '# ðŸ˜„âœŒ'
                ]
            }
        ]
    };
};
