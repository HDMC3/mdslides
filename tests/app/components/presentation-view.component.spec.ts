import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { NbEvaIconsModule } from '@nebular/eva-icons';
import { NbIconModule, NbThemeModule } from '@nebular/theme';
import { MarpitService } from 'src/app/core/services/marpit.service';
import { PresentationService } from 'src/app/core/services/presentation.service';

import { PresentationViewComponent } from '../../../src/app/components/presentation-view/presentation-view.component';

describe('PresentationViewComponent', () => {
    let component: PresentationViewComponent;
    let fixture: ComponentFixture<PresentationViewComponent>;
    let compiled: HTMLElement;
    beforeEach(async() => {
        await TestBed.configureTestingModule({
            declarations: [PresentationViewComponent],
            imports: [
                NbThemeModule.forRoot({ name: 'default' }),
                RouterTestingModule,
                NbIconModule,
                NbEvaIconsModule
            ],
            providers: [
                PresentationService,
                MarpitService
            ]
        })
            .compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(PresentationViewComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
        compiled = fixture.nativeElement;
    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(compiled).toMatchSnapshot();
    });
});
