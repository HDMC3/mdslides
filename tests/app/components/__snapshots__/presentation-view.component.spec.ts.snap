// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PresentationViewComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="view-container"
  >
    <div
      class="slides-track"
      style="transform: translateX(0px);"
    >
      
      <div
        class="slide"
        style="--scale: 1;"
      />
    </div>
    <button
      class="next"
    >
      <nb-icon
        icon="arrow-forward"
        style="color: var(--color-primary-400); width: 50px; height: 50px;"
      >
        <svg
          class="eva eva-arrow-forward"
          fill="currentColor"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            data-name="Layer 2"
          >
            <g
              data-name="arrow-forward"
            >
              <rect
                height="24"
                opacity="0"
                transform="rotate(-90 12 12)"
                width="24"
              />
              <path
                d="M5 13h11.86l-3.63 4.36a1 1 0 0 0 1.54 1.28l5-6a1.19 1.19 0 0 0 .09-.15c0-.05.05-.08.07-.13A1 1 0 0 0 20 12a1 1 0 0 0-.07-.36c0-.05-.05-.08-.07-.13a1.19 1.19 0 0 0-.09-.15l-5-6A1 1 0 0 0 14 5a1 1 0 0 0-.64.23 1 1 0 0 0-.13 1.41L16.86 11H5a1 1 0 0 0 0 2z"
              />
            </g>
          </g>
        </svg>
      </nb-icon>
    </button>
    <button
      class="previous"
    >
      <nb-icon
        icon="arrow-back"
        style="color: var(--color-primary-400); width: 50px; height: 50px;"
      >
        <svg
          class="eva eva-arrow-back"
          fill="currentColor"
          height="100%"
          viewBox="0 0 24 24"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            data-name="Layer 2"
          >
            <g
              data-name="arrow-back"
            >
              <rect
                height="24"
                opacity="0"
                transform="rotate(90 12 12)"
                width="24"
              />
              <path
                d="M19 11H7.14l3.63-4.36a1 1 0 1 0-1.54-1.28l-5 6a1.19 1.19 0 0 0-.09.15c0 .05 0 .08-.07.13A1 1 0 0 0 4 12a1 1 0 0 0 .07.36c0 .05 0 .08.07.13a1.19 1.19 0 0 0 .09.15l5 6A1 1 0 0 0 10 19a1 1 0 0 0 .64-.23 1 1 0 0 0 .13-1.41L7.14 13H19a1 1 0 0 0 0-2z"
              />
            </g>
          </g>
        </svg>
      </nb-icon>
    </button>
  </div>
</div>
`;
